k2/csrc/fsa_algo.cc

```diff
@@ -275,6 +275,111 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   return is_acyclic;
 }
 
+void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
+                         std::vector<std::vector<int32_t>> *arc_derivs) {
+  CHECK_EQ(a.weight_type, kMaxWeight);
+  CHECK_GT(beam, 0);
+  CHECK_NOTNULL(b);
+  CHECK_NOTNULL(arc_derivs);
+  b->arc_indexes.clear();
+  b->arcs.clear();
+  arc_derivs->clear();
+
+  const auto &fsa = a.fsa;
+  if (IsEmpty(fsa)) return;
+  int32_t num_states_a = fsa.NumStates();
+  int32_t final_state = fsa.FinalState();
+  const auto &arcs_a = fsa.arcs;
+  const float *arc_weights_a = a.arc_weights;
+
+  // identify all states that should be kept
+  std::vector<char> non_eps_in(num_states_a, 0);
+  non_eps_in[0] = 1;
+  using WeightsPair = std::pair<double, std::vector<int32_t>>;
+  // (label, dest_state) -> `sum` of weights along all paths from current state
```

Comment:
clarify that sum here == max.  probably just say max would be clearer.

Reply:
You don't need to sum over a specific label here.  We are just removing epsilons, not determinizing.

Reply:
OK, will change sum to max.

RE `sum over label`, if there is two epsilon arcs from `a->b`, seems we need to do `sum(max)` here, right?

Reply:
yes, for epsilons, but doing that for non-epsilon labels is unnecessary.

Reply:
.. calling epsilons a label in this context would be a little strange, given that epsilon means "no label here".